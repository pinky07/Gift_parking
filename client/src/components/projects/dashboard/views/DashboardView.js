import React from 'react';
import Toast from 'grommet/components/Toast';

import DashboardWithCriticalError from './viewComponents/DashboardWithCriticalError';
import LoadingView from './viewComponents/LoadingView';
import DashboardWithNoCycleSnaps from './viewComponents/DashboardWithNoCycleSnaps';
import DashboardWithCycleSnaps from './viewComponents/DashboardWithCycleSnaps';

import AddCycleSnapController from '../../addCycleSnap/AddCycleSnapController';
import AddReleaseSnapController from "../../addReleaseSnap/AddReleaseSnapController";
import AddIncidentsReportController from "../../addIncidentsReport/AddIncidentsReportController";

const DashboardView = ({viewModel}) => {
  const {errorMessage} = viewModel;

  if (errorMessage)
    return (<DashboardWithCriticalError errorMessage={errorMessage}/>);
  else {
    const {name} = viewModel.project;
    if (name) {
      const elements = determineElements(viewModel);
      const {cycleSnaps} = viewModel.project;
      if (cycleSnaps.length > 0) {
        return (<DashboardWithCycleSnaps viewModel={viewModel} elements={elements}/>);
      }
      else
        return (<DashboardWithNoCycleSnaps viewModel={viewModel} elements={elements}/>);
    }
    else
      return (<LoadingView/>);
  }
}

/* Determines the UI elements that will be shown: 
  - success notification after adding a cycle snap 
  - failure notification when adding a cycle snap 
  - the modal dialog for adding a cycle snap 
*/
const determineElements = (viewModel) => {
  let elements = {
    successNotification: undefined,
    failureNotification: undefined,
    addCycleSnapLayer: undefined,
    addReleaseSnapLayer: undefined,
    addIncidentsReportLayer: undefined
  };

  const {successNotificationOnAdd} = viewModel;
  if (successNotificationOnAdd)
    elements.successNotification = <Toast status='ok'>{successNotificationOnAdd}</Toast>;

  const {failureNotificationOnAdd} = viewModel;
  if (failureNotificationOnAdd)
    elements.failureNotification = <Toast status='critical'>{failureNotificationOnAdd}</Toast>;

  const {dashboardCallbacks} = viewModel;
  const {addCycleSnap} = viewModel;
  const {addReleaseSnap} = viewModel;
  const {addIncidentsReport} = viewModel;
  
  if (addCycleSnap) {
    elements.addCycleSnapLayer = <AddCycleSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddCycleSnapSubmit}/>;
  }
  
  if (addReleaseSnap) {
    elements.addReleaseSnapLayer = <AddReleaseSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddReleaseSnapSubmit}/>;
  }

  if (addIncidentsReport) {
    elements.addIncidentsReportLayer = <AddIncidentsReportController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddIncidentsReportSubmit}/>;
  }

  return elements;
}

export default DashboardView;