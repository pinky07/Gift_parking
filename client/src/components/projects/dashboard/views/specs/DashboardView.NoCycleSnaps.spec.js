import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import DashboardView from '../DashboardView';

describe('DashboardView when project does not have cycle snaps', () => {

    let viewModel;
    let actual;

    it('should render dashboard with no cycle snaps', () => {
        const onRequestAddCycleSnap = jest.fn();
        const onAddCycleSnapSubmit = jest.fn();
        const onAddCycleSnapCancel = jest.fn();

        viewModel = {
            projectId: '2',
            project: getProjectWithNoCycleSnaps(),
            dashboardCallbacks: {
                onRequestAddCycleSnap: onRequestAddCycleSnap,
                onAddCycleSnapSubmit: onAddCycleSnapSubmit,
                onClose: onAddCycleSnapCancel
            },

            errorMessage: '',
            addCycleSnap: false,
            successNotificationOnAdd: undefined,
            failureNotificationOnAdd: undefined
        };

        const renderer = ShallowRenderer.createRenderer();
        actual = renderer.render(<DashboardView viewModel={viewModel} />);

        expect(actual).toMatchSnapshot();
    });

    function getProjectWithNoCycleSnaps() {
        // Test data from http://localhost:8080/api/v1/projects/2/dashboard 
        return { "name": "BAAM", "cycleSnaps": [] };
    }

});