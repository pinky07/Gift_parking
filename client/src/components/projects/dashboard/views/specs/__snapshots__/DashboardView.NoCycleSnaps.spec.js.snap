// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardView when project does not have cycle snaps should render dashboard with no cycle snaps 1`] = `
<DashboardWithNoCycleSnaps
  elements={
    Object {
      "addCycleSnapLayer": undefined,
      "addIncidentsReportLayer": undefined,
      "addReleaseSnapLayer": undefined,
      "failureNotification": undefined,
      "successNotification": undefined,
    }
  }
  viewModel={
    Object {
      "addCycleSnap": false,
      "dashboardCallbacks": Object {
        "onAddCycleSnapSubmit": [Function],
        "onClose": [Function],
        "onRequestAddCycleSnap": [Function],
      },
      "errorMessage": "",
      "failureNotificationOnAdd": undefined,
      "project": Object {
        "cycleSnaps": Array [],
        "name": "BAAM",
      },
      "projectId": "2",
      "successNotificationOnAdd": undefined,
    }
  }
/>
`;
